Return jacobian_sparsity() when sparse && !jac

akashkgarg opened the pull request:
Return jacobian_sparsity() when sparse && !jac. Resolves #871 

At the time of writing this, I realized there's another open PR that addresses this ticket: 
https://github.com/SciML/ModelingToolkit.jl/pull/966

ChrisRackauckas commented:
I think this is a better solution because it doesn't double compute the sparsity pattern when it doesn't need to. But this doesn't do the last little bit: it needs to upconvert from the `SparseMatrixCSC{Bool}` to the `eltype(u0)` during the construction of the ODEProblem.

akashkgarg commented:
Ah okay, so we want to add something like `convert.(eltype(u0), jacobian_sparsity(sys))`? I had assumed that if a user wanted the sparsity "pattern" they would expect it in Bool format, but that probably doesn't compose well.

ChrisRackauckas commented:
The differential equation solver just makes a copy of the `jac_prototype`, so it will error when it tries to fill it with floats. This is why it's called "prototype": it's a prototype of what it's going to copy, not just the pattern itself. Though in theory the differential equation solver could do the conversion in its setup phase: that would be another way to fix it.

YingboMa commented:
Yeah, so something like `similar(f.jac_prototype, uEltype)`.

akashkgarg commented:
Okay, fixed the typing issue in latest push.

YingboMa commented:
No, this's not right. It will error when a user just call ODEFunction.

akashkgarg commented:
@YingboMa Could you expand on your comment? I tried running tests in ModelingToolkit and all of them pass except for some in latexify.jl, but I think those are unrelated to this change? It's likely I'm missing something or we need to add another test that captures the failure case you're thinking of?

ChrisRackauckas commented on src/systems/diffeqs/abstractodesystem.jl:
```diff
@@ -215,12 +215,13 @@ function DiffEqBase.ODEFunction{iip}(sys::AbstractODESystem, dvs = states(sys),
         end
     end
 
+    uElType = eltype(u0)
```
I think this isn't safe for the `u0 == nothing` case.

ChrisRackauckas commented:
>No, this's not right. It will error when a user just call ODEFunction.

I don't think it will?

ChrisRackauckas commented:
This is missing a test.

YingboMa commented:
> I don't think it will?

It will fail when u0 isn't passed into ODEFunction.

ChrisRackauckas commented:
Yes, that's what I meant by "I think this isn't safe for the u0 == nothing case.". Is the solution here to add a convert in each of the solvers? Use `SetField.jl` in ODEProblem? Etc.

YingboMa commented:
Handling this in the solvers' level would be great. @akashkgarg do you want to file a PR to OrdinaryDiffEq, StochasticDiffEq, and Sundials?

akashkgarg commented:
Okay, I updated the code to explicitly check for the case when `u0` is `nothing` and also added some tests. I can take a stab at modifying the solvers as well.

YingboMa commented:
It might be easier to overload `get_concrete_problem` in https://github.com/SciML/DiffEqBase.jl/blob/7824bd8cec9d98ff6122acef99fe80d7dee8dddb/src/solve.jl#L138-L192

Note that not all `jac_prototype`s have `similar` defined, so you need to check `isa AbstractArray` before calling `similar(jac_prototype, eltype(u0))` and such, e.g.
```julia
julia> using DiffEqOperators, LinearAlgebra

julia> A = 0.01*rand(3, 3)
3×3 Matrix{Float64}:
 0.00423074   0.00891582  0.00839771
 0.00382897   0.00291319  0.00918495
 0.000318021  0.00235121  0.0080333

julia> rn = (du, u, p, t) -> begin
           mul!(du, A, u)
       end
#1 (generic function with 1 method)

julia> u0 = rand(3)
3-element Vector{Float64}:
 0.6075074514491132
 0.8949122350449714
 0.8993517833428093

julia> A = JacVecOperator{Float64}(rn, u0; autodiff=false);

julia> similar(A, Int)
ERROR: MethodError: no method matching similar(::JacVecOperator{Float64, var"#1#2", Vector{Float64}, Vector{Float64}, Vector{Float64}, Any, Any, Bool}, ::Type{Int64})
Closest candidates are:
  similar(::Union{Adjoint{T, var"#s832"}, Transpose{T, var"#s832"}} where {T, var"#s832"<:(AbstractVector{T} where T)}, ::Type{T}) where T at /Users/scheme/.bin/julia-release/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:230
  similar(::Union{Adjoint{T, S}, Transpose{T, S}} where {T, S}, ::Type{T}) where T at /Users/scheme/.bin/julia-release/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:234
  similar(::Union{Adjoint{T, S}, Transpose{T, S}} where {T, S}, ::Type{T}, ::Tuple{Vararg{Int64, N}}) where {T, N} at /Users/scheme/.bin/julia-release/usr/share/julia/stdlib/v1.6/LinearAlgebra/src/adjtrans.jl:235
  ...
Stacktrace:
 [1] top-level scope
   @ REPL[6]:1

julia> A isa AbstractArray
false
```

ChrisRackauckas commented:
Yes, adding that to `get_concrete_problem` and using SetField.jl could be the right way to do this.

akashkgarg commented:
I took a stab at this here: SciML/DiffEqBase.jl#657 . Not sure if it catches all the cases, but as far as I can tell, promote_f gets called before any solve that happens related to ODEFunction. Let me know if something like this is what you had in mind or if there might be a better solution.

ChrisRackauckas commented:
There seems to still be a test failure here. Could you take a look at it?

https://github.com/SciML/ModelingToolkit.jl/pull/968/checks?check_run_id=2362637867#step:6:370

akashkgarg commented:
Weird. These tests pass locally for me. Probably some weird tolerance checking with floating point.

ChrisRackauckas commented:
Increasing the tolerance a bit is probably fine.

YingboMa commented on test/jacobiansparsity.jl:
```diff
@@ -0,0 +1,73 @@
+
+# test sparse jacobian pattern only.
+prob = ODEProblem(sys, u0, (0, 11.5), sparse=true, jac=false)
+@test isapprox(similar(Symbolics.jacobian_sparsity(map(x->x.rhs, equations(sys)), states(sys)),Float64), prob.f.jac_prototype; atol=1.e-4)
```
What does this do? Are you comparing uninitialize memory with another piece of uninitialized memory?

ChrisRackauckas replied:
Yes, for reference `similar` is uninitialized memory, so its values are essentially random, which is why this check fails. I think you should check `zero(A)` instead of `similar(A)`, or use the `(I,J,V) = findnz(A)` and compare the `I` and `J`.

YingboMa replied:
Since it's `jac_prototype`, we should only check the sparsity pattern, but not the values.

akashkgarg replied:
Hmm, I did not realize that `similar` returns an uninitialized array. Shouldn't we be using `convert` in this case? My understanding is that you get the sparsity pattern using`jacobian_sparsity(sys)`, which is an array of `bool`, but we want to make sure the type matches that of `u0`. Seems like `convert` is what we need over `similar`?

ChrisRackauckas replied:
`convert.(Float64,x)` would guarantee it to be 1's and 0's, and so would `float.`, though for a test directly checking `I` and `J` might be safer.

YingboMa reviewed (changes requested):
The tests are wrong.

akashkgarg commented:
It seems to me that for `jac_prototype` all we care about is the sparsity pattern? If so, is there ever a reason to return `similar(get_jac(sys)[],uElType))`? Can we simply just always return `similar(jacobian_sparsity(sys),uElType)` when `sparse = true`?

YingboMa commented:
We should not use `get_jac(sys)` but just call `J = calculate_jacobian(sys)`. And to get the sparsity pattern, we just do `similar(J, uElType)`. Since if we have the Jacobian already, we might as well just generate the prototype from the symbolic Jacobian.

ChrisRackauckas commented:
Awesome, that one had a bit more too it than I had anticipated.
